src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,154 @@
+using namespace bsoncxx;
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+/** @private */
+struct SamplingLoader::PhaseConfig {
+    PhaseConfig(PhaseContext& context,
+                mongocxx::pool::entry& client,
+                uint threadId,
+                size_t totalThreads,
+                ActorId id)
+        : database{(*client)[context["Database"].to<std::string>()]},
+          collection{database[context["Collection"].to<std::string>()]},
+          // The next line uses integer division for non multithreaded configurations.
+          // The Remainder is accounted for below.
+          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()},
+          sampleSize{
+              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
+          threadId(threadId) {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    uint threadId;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(info) << "Beginning to run SamplingLoader";
+            // Read the sample size documents into a vector.
+            mongocxx::pipeline pipe;
+            pipe.sample(config->sampleSize);
+            pipe.project(make_document(kvp("_id", 0)));
+            auto cursor = config->collection.aggregate(pipe, mongocxx::options::aggregate{});
+            const auto sampleDocs =
+                std::vector<bsoncxx::document::value>(cursor.begin(), cursor.end());
+
+            if (sampleDocs.empty()) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException("Collection has no documents"));
+            }
+            if (sampleDocs.size() < config->sampleSize) {
+                BOOST_THROW_EXCEPTION(
+                    InvalidConfigurationException("Collection smaller than sample size"));
+            }
+
+            // Maintain an index into the sample that's used for the insert batches so we
+            // insert roughly the same number of copies for each sample.
+            size_t sampleIdx = 0;
+
+            auto batchOfDocs = std::vector<bsoncxx::document::view_or_value>{};
+            auto totalOpCtx = _totalBulkLoad.start();
+            for (size_t batch = 0; batch < config->numBatches; ++batch) {
+                batchOfDocs.clear();
+                for (size_t i = 0; i < config->insertBatchSize; ++i) {
+                    batchOfDocs.push_back(sampleDocs[sampleIdx].view());
+                    sampleIdx = (sampleIdx + 1) % sampleDocs.size();
+                }
+
+                // Now do the insert.
+                auto individualOpCtx = _individualBulkLoad.start();
+                try {
+                    auto result = config->collection.insert_many(std::move(batchOfDocs));
```

Comment:
Is it safe to use a `vector` after it's been `move`d from? I think you might need to re-init `batchOfDocs` after doing this.

Reply:
this is not safe anymore if reusing the vector, right? why collection.insert_many wants a copy? (if it insists on using a copy might as well create the vector each time as you were doing before, apologies for the distraction)

Reply:
Oh shoot, yeah if we're trying to re-use the same memory then this backfires. Undone!

Reply:
I'd still reserve the total capacity of the vector before pushing new docs into as for bigger batch sizes it might save a few reallocations